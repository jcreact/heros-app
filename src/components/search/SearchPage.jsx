import React from 'react';
import queryString from 'query-string';
import { makeStyles } from '@material-ui/core/styles';
import {
    Box,
    Divider,
    Fade,
    Grid,
    TextField,
    Typography,
} from '@material-ui/core';
import { useState } from 'react';

import { HeroListItem } from '../heros/HeroListItem';
import { getHerosByFilter } from '../../selectors/getHeroByFilter';
import { useLocation } from 'react-router-dom';
import { useMemo } from 'react';

const useStyles = makeStyles((theme) => ({
    divider: {
        marginTop: theme.spacing(2),
        marginBottom: theme.spacing(2),
    },
}));

export const SearchPage = ({ history }) => {
    const cls = useStyles();

    const { search: query } = useLocation();
    const { q = '' } = queryString.parse(query);

    const [search, setSearch] = useState(q);

    const heros = useMemo(() => getHerosByFilter(q), [q]);

    const handleChange = (value) => {
        setSearch(value);
    };

    const handleSumit = (evt) => {
        evt.preventDefault();
        history.push(`?q=${search}`);
    };

    return (
        <Box mt={4}>
            <form onSubmit={handleSumit}>
                <TextField
                    label="Buscador"
                    placeholder="Nombre de hÃ©roe..."
                    fullWidth
                    InputLabelProps={{
                        shrink: true,
                    }}
                    value={search}
                    variant="outlined"
                    autoComplete="off"
                    onChange={(e) => handleChange(e.target.value)}
                />
            </form>
            {heros.length > 0 && (
                <Fade in timeout={500}>
                    <Box mt={2}>
                        <Typography variant="h5">Resultados</Typography>
                        <Divider className={cls.divider} />
                        <Grid container spacing={2}>
                            {heros.map((hero) => (
                                <Grid item xs={12} sm={4} md={3} key={hero.id}>
                                    <HeroListItem {...hero} />
                                </Grid>
                            ))}
                        </Grid>
                    </Box>
                </Fade>
            )}
        </Box>
    );
};
