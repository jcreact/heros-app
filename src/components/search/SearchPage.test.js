import { MemoryRouter, Route } from 'react-router-dom';
import { mount } from 'enzyme';
import { SearchPage } from './SearchPage';
import { act } from 'react-dom/test-utils';

describe(`Pruebas en el componente 'SearchPage.jsx'.`, () => {

    const history = {
        push: jest.fn()
    };

    test('Debe renderizar el componente correctamente.', () => {
        const component = mount(
            <MemoryRouter initialEntries={['/search']}>
                <Route path="/search" component={() => <SearchPage history={history} />} />
            </MemoryRouter>
        );

        expect(component).toMatchSnapshot();
        expect(component.find('input').exists()).toBeTruthy();
        expect(component.find('label').exists()).toBeTruthy();
        expect(component.find('h5').exists()).toBe(false);
    });


    test(`Debe a 'Flash' y el valor del input debe ser 'flash'.`, () => {
        const component = mount(
            <MemoryRouter initialEntries={['/search?q=flash']}>
                <Route path="/search" component={() => <SearchPage history={history} />} />
            </MemoryRouter>
        );

        expect(component).toMatchSnapshot();
        expect(component.find('input').prop('value')).toBe('flash');
        expect(component.find('h5').text()).toBe('Resultados');
        expect(component.find('div.MuiGrid-item').exists()).toBeTruthy();
        expect(component.find('div.MuiAvatar-root').text()).toBe('F');
        expect(component.find('img[title="Flash"]').exists()).toBeTruthy();
    });


    test(`No bebe de mostrar ningún resultado si q es 'hero-mock'.`, () => {
        const component = mount(
            <MemoryRouter initialEntries={['/search?q=hero-mock']}>
                <Route path="/search" component={() => <SearchPage history={history} />} />
            </MemoryRouter>
        );

        expect(component).toMatchSnapshot();
        expect(component.find('input').prop('value')).toBe('hero-mock');
        expect(component.find('h5').text()).toBe('No se encontraron coincidencias...');
    });

    test(`Debe realizar la búsqueda para batmap.`, () => {

        const component = mount(
            <MemoryRouter initialEntries={['/search']}>
                <Route path="/search" component={() => <SearchPage history={history} />} />
            </MemoryRouter>
        );

        // act(() => {
        //     component.find('input').prop('onChange')({ target: { value: 'batman' } });
        // });
        // component.update();

        component.find('input').simulate('change', { target: { value: 'batman' } });
        component.find('form').prop('onSubmit')({
            preventDefault() { }
        });

        expect(component.find('input').prop('value')).toBe('batman');
        expect(history.push).toHaveBeenCalledWith('?q=batman');
    });

});
