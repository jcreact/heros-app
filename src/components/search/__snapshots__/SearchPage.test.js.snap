// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el componente 'SearchPage.jsx'. Debe a 'Flash' y el valor del input debe ser 'flash'. 1`] = `
<div
  className="MuiBox-root MuiBox-root-7"
>
  <form
    onSubmit={[Function]}
  >
    <div
      className="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
    >
      <label
        className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-filled"
        data-shrink={true}
      >
        Buscador
      </label>
      <div
        className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl"
        onClick={[Function]}
      >
        <input
          aria-invalid={false}
          autoComplete="off"
          autoFocus={false}
          className="MuiInputBase-input MuiOutlinedInput-input"
          disabled={false}
          onAnimationStart={[Function]}
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          placeholder="Nombre de héroe..."
          required={false}
          type="text"
          value="flash"
        />
        <fieldset
          aria-hidden={true}
          className="PrivateNotchedOutline-root-3 MuiOutlinedInput-notchedOutline"
        >
          <legend
            className="PrivateNotchedOutline-legendLabelled-5 PrivateNotchedOutline-legendNotched-6"
          >
            <span>
              Buscador
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
  </form>
  <div
    className="MuiBox-root MuiBox-root-8"
    style={
      Object {
        "opacity": 1,
        "visibility": undefined,
      }
    }
  >
    <h5
      className="MuiTypography-root MuiTypography-h5"
    >
      Resultados
    </h5>
    <hr
      className="MuiDivider-root makeStyles-divider-1"
    />
    <div
      className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2"
    >
      <div
        className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-4 MuiGrid-grid-md-3"
      >
        <div
          className="MuiPaper-root MuiCard-root MuiPaper-outlined MuiPaper-rounded"
        >
          <div
            className="MuiCardHeader-root makeStyles-cardTitle-11"
          >
            <div
              className="MuiCardHeader-avatar"
            >
              <div
                className="MuiAvatar-root MuiAvatar-circle makeStyles-avatarDC-10 MuiAvatar-colorDefault"
              >
                F
              </div>
            </div>
            <div
              className="MuiCardHeader-content"
            >
              <span
                className="MuiTypography-root MuiCardHeader-title MuiTypography-body2 MuiTypography-displayBlock"
              >
                Flash
              </span>
              <span
                className="MuiTypography-root MuiCardHeader-subheader MuiTypography-body2 MuiTypography-colorTextSecondary MuiTypography-displayBlock"
              >
                Jay Garrick
              </span>
            </div>
            <div
              className="MuiCardHeader-action"
            >
              <button
                className="MuiButtonBase-root MuiIconButton-root"
                disabled={false}
                onBlur={[Function]}
                onDragLeave={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                onKeyUp={[Function]}
                onMouseDown={[Function]}
                onMouseLeave={[Function]}
                onMouseUp={[Function]}
                onTouchEnd={[Function]}
                onTouchMove={[Function]}
                onTouchStart={[Function]}
                tabIndex={0}
                type="button"
              >
                <span
                  className="MuiIconButton-label"
                >
                  <svg
                    aria-hidden={true}
                    className="MuiSvgIcon-root"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M12 8c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2zm0 2c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm0 6c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2z"
                    />
                  </svg>
                </span>
                <span
                  className="MuiTouchRipple-root"
                />
              </button>
            </div>
          </div>
          <img
            className="MuiCardMedia-root MuiCardMedia-media MuiCardMedia-img"
            src="."
            title="Flash"
          />
          <div
            className="MuiCardContent-root"
          >
            <p
              className="MuiTypography-root MuiTypography-body1"
            >
              Flash Comics #1
            </p>
            <p
              className="MuiTypography-root MuiTypography-body1"
            >
              Jay Garrick, Barry Allen, Wally West, Bart Allen
            </p>
          </div>
          <div
            className="MuiCardActions-root MuiCardActions-spacing"
          >
            <a
              aria-current={null}
              aria-disabled={false}
              aria-label="learn more"
              className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary"
              href="/hero/dc-flash"
              onBlur={[Function]}
              onClick={[Function]}
              onDragLeave={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              role="button"
              tabIndex={0}
            >
              <span
                className="MuiButton-label"
              >
                ver más
              </span>
              <span
                className="MuiTouchRipple-root"
              />
            </a>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Pruebas en el componente 'SearchPage.jsx'. Debe renderizar el componente correctamente. 1`] = `
<div
  className="MuiBox-root MuiBox-root-2"
>
  <form
    onSubmit={[Function]}
  >
    <div
      className="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
    >
      <label
        className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined"
        data-shrink={true}
      >
        Buscador
      </label>
      <div
        className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl"
        onClick={[Function]}
      >
        <input
          aria-invalid={false}
          autoComplete="off"
          autoFocus={false}
          className="MuiInputBase-input MuiOutlinedInput-input"
          disabled={false}
          onAnimationStart={[Function]}
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          placeholder="Nombre de héroe..."
          required={false}
          type="text"
          value=""
        />
        <fieldset
          aria-hidden={true}
          className="PrivateNotchedOutline-root-3 MuiOutlinedInput-notchedOutline"
        >
          <legend
            className="PrivateNotchedOutline-legendLabelled-5 PrivateNotchedOutline-legendNotched-6"
          >
            <span>
              Buscador
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
  </form>
</div>
`;

exports[`Pruebas en el componente 'SearchPage.jsx'. No bebe de mostrar ningún resultado si q es 'hero-mock'. 1`] = `
<div
  className="MuiBox-root MuiBox-root-12"
>
  <form
    onSubmit={[Function]}
  >
    <div
      className="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
    >
      <label
        className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-filled"
        data-shrink={true}
      >
        Buscador
      </label>
      <div
        className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl"
        onClick={[Function]}
      >
        <input
          aria-invalid={false}
          autoComplete="off"
          autoFocus={false}
          className="MuiInputBase-input MuiOutlinedInput-input"
          disabled={false}
          onAnimationStart={[Function]}
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          placeholder="Nombre de héroe..."
          required={false}
          type="text"
          value="hero-mock"
        />
        <fieldset
          aria-hidden={true}
          className="PrivateNotchedOutline-root-3 MuiOutlinedInput-notchedOutline"
        >
          <legend
            className="PrivateNotchedOutline-legendLabelled-5 PrivateNotchedOutline-legendNotched-6"
          >
            <span>
              Buscador
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
  </form>
  <div
    className="MuiBox-root MuiBox-root-13"
    style={
      Object {
        "opacity": 1,
        "visibility": undefined,
      }
    }
  >
    <h5
      className="MuiTypography-root MuiTypography-h5 MuiTypography-colorError"
    >
      No se encontraron coincidencias...
    </h5>
  </div>
</div>
`;
